llvm/lib/Analysis/DXILResource.cpp

```diff
@@ -823,8 +827,153 @@ DXILBindingMap::findByUse(const Value *Key) const {
 
 //===----------------------------------------------------------------------===//
 
+static bool isUpdateCounterIntrinsic(Function &F) {
+  return F.getIntrinsicID() == Intrinsic::dx_resource_updatecounter;
+}
+
+void DXILResourceCounterDirectionMap::populate(Module &M, DXILBindingMap &DBM) {
+  std::vector<std::tuple<dxil::ResourceBindingInfo, ResourceCounterDirection,
+                         const Function *, const CallInst *>>
+      DiagCounterDirs;
```

Comment:
Better to use `SmallVector` here rather than `std::vector`. Also I'm surprised this is holding a `dxil::ResourceBindingInfo` by value rather than reference - more on that below.